redash/models.py

```diff
@@ -1390,26 +1390,19 @@ def search(cls, org, groups_ids, user_id, search_term):
 
     @classmethod
     def all_tags(cls, org, user):
+        dashboards = cls.all(org, user.group_ids, user.id)
+
         tag_column = func.unnest(cls.tags).label('tag')
         usage_count = func.count(1).label('usage_count')
 
         query = (
-            db.session.query(tag_column, usage_count)
-            .outerjoin(Widget)
-            .outerjoin(Visualization)
-            .outerjoin(Query)
-            .outerjoin(DataSourceGroup, Query.data_source_id == DataSourceGroup.data_source_id)
-            .filter(
-                Dashboard.is_archived == False,
-                (DataSourceGroup.group_id.in_(user.group_ids) |
-                 (Dashboard.user_id == user.id) |
-                 ((Widget.dashboard != None) & (Widget.visualization == None))),
-                Dashboard.org == org)
-            .group_by(tag_column))
-
-        query = query.filter(or_(Dashboard.user_id == user.id, Dashboard.is_draft == False))
-
-        return query.order_by(usage_count.desc())
+            db.session
+            .query(tag_column, usage_count)
+            .group_by(tag_column)
+            .filter(Dashboard.id.in_(dashboards.options(load_only('id'))))
+            .order_by(usage_count.desc())
+        )
+        return query
```

Comment:
Can I bother you with adding a test case for this bug & method?

Reply:
No bother :)